Add os.release facts on FreeBSD

smortex opened the pull request:
This PR provide support for Facter 3.x `os.release` facts on FreeBSD to Facter 4:

Before:
```sh-session
romain@zappy ~/Projects/facter-ng % bundle exec bin/facter os        
cat: /etc/release: No such file or directory
[2020-04-29 18:51:23.841325 ] ERROR Facter::Resolvers::SolarisRelease - Could not build release fact because of missing or empty file /etc/release 
cat: /etc/release: No such file or directory
[2020-04-29 18:51:23.842256 ] ERROR Facter::Resolvers::SolarisRelease - Could not build release fact because of missing or empty file /etc/release 
cat: /etc/release: No such file or directory
[2020-04-29 18:51:23.843051 ] ERROR Facter::Resolvers::SolarisRelease - Could not build release fact because of missing or empty file /etc/release 
{
  architecture => "amd64",
  family => "FreeBSD",
  hardware => "amd64",
  name => "FreeBSD",
  release => {
    full => null,
    major => null,
    minor => null
  }
}
```

After:
```sh-session
romain@zappy ~/Projects/facter-ng % bundle exec bin/facter os      
{
  architecture => "amd64",
  family => "FreeBSD",
  hardware => "amd64",
  name => "FreeBSD",
  release => {
    branch => "RELEASE-p4",
    full => "12.1-RELEASE-p4",
    major => "12",
    minor => "1",
    patchlevel => "4"
  }
}
```

Compare to:

```sh-session
romain@zappy ~/Projects/facter-ng % facter os
{
  architecture => "amd64",
  family => "FreeBSD",
  hardware => "amd64",
  name => "FreeBSD",
  release => {
    branch => "RELEASE-p4",
    full => "12.1-RELEASE-p4",
    major => "12",
    minor => "1",
    patchlevel => "4"
  }
}
```

smortex committed:
(maint) Add os.release facts on FreeBSD

puppetlabs-jenkins commented:
Can one of the admins verify this patch?

smortex commented on os_hierarchy.json:
```diff
@@ -24,7 +24,11 @@
+        "Bsd": [
+          "Freebsd"
+        ]
+      }
```
FreeBSD is definitively a descendant of BSD, yet BSD is not a decedent of Solaris (amusingly, [early version of Sun OS (which became Solaris at some point) was based on BSD](https://en.wikipedia.org/wiki/Solaris_(operating_system)#History)).

For now, this is convenient because Solaris facts and resolvers provide ZFS related facts which are working on FreeBSD, however, ZFS on Linux is a thing and some refactoring will be necessary at some point.  Don't know how / where to log this so that it can be tracked…

smortex replied:
And maybe it is as simple as renaming the Solaris ZFS resolver to make it obvious it is not Solaris specific and use similar facts for FreeBSD, Solaris and maybe Linux to gather these facts…  So basically the same code facts multiple times, but maybe it's not a problem?

IrimieBogdan replied:
@smortex I talked with @oanatmaria and maybe we can have something like
```
...
  {
    "Unix": [
      "Solaris",
      {
        "Bsd": [
          "Freebsd"
        ]
      }

    ]
  },
...
```
Both `Solaris` and `Bsd` will inherit from `Unix`. This is based on the diagram from https://upload.wikimedia.org/wikipedia/commons/7/77/Unix_history-simple.svg

oanatmaria commented on lib/resolvers/freebsd/freebsd_version_resolver.rb:
```diff
@@ -0,0 +1,36 @@
+          def build_fact_list(output)
+            freebsd_version_results = output.split("\n")
+
+            @fact_list[:installed_kernel]   = freebsd_version_results[0].strip
+            @fact_list[:running_kernel]     = freebsd_version_results[1].strip
+            @fact_list[:installed_userland] = freebsd_version_results[2].strip
```
.strip will raise NoMethodError if output is empty string or if freebsd_version_results[0/1/2] is nil

smortex replied:
`freebsd-version -kru` is expected to return exactly 3 lines (each flag produce a line), so I do not thing this is a problem for us here.  Yet, some FreeBSD forks might be detected as "FreeBSD" while not providing `freebsd-update`, so it's worth being sure that the command succeeded.

igalic replied:
HardenedBSD doesn't.

smortex replied:
@igalic did you have the opportunity to try it on HardenedBSD?  I need to allocate some time to dig in this project…

igalic replied:
unfortunately, my current priority
https://github.com/canonical/cloud-init/pull/363 is a few yak stacks away from testing HardenedBSD

oanatmaria commented on lib/facts/freebsd/os/release.rb:
```diff
@@ -0,0 +1,41 @@
+
+        def build_release_hash_from_version(version_string)
+          version, branch_value = version_string.split('-', 2)
+          major_value, minor_value = version.split('.')
+          patchlevel_value = branch_value.split('-p')[1]
```
if version_string is nil or empty string it will raise NoMethodError

smortex replied:
D'oh, and it's implicitly broken…  heavy refactoring to make Rubocop happier made me break the code :unamused: 

Will check tomorrow, thanks for the review!

oanatmaria commented on lib/resolvers/freebsd/freebsd_version_resolver.rb:
```diff
@@ -0,0 +1,36 @@
+          end
+
+          def freebsd_version_system_call(fact_name)
+            output, _status = Open3.capture2('/bin/freebsd-version -kru')
```
prefer capture3 because it catches stderr also, otherwise stderr will be displayed if command fails

smortex replied:
Makes sense

smortex committed:
(maint) Improve freebsd-version failure management

This program is always supposed to be found and return the expected
output on FreeBSD but make sure it runs successfully, catpure stderr in
case it becomes used at some point, and do not raise exceptions if the
resolver fails.

oanatmaria commented on lib/facts/freebsd/os/release.rb:
```diff
@@ -0,0 +1,43 @@
+        def call_the_resolver
+          installed_userland = Facter::Resolvers::Freebsd::FreebsdVersion.resolve(:installed_userland)
+
+          return [] if installed_userland.nil?
```
return Facter::ResolvedFact.new(FACT_NAME, nil) if !installed_userland || installed_userland.empty?

smortex replied:
Updated, thanks!

smortex committed:
(maint) Improve failing fact resolution

If the resolver fails to gather the fact, return an empty resolved fact
instead of an empty array.

smortex committed:
(maint) Make rubocop happy

spec/facter/resolvers/utils/windows/network_utils_spec.rb:37:7: C: RSpec/EmptyHook: Empty hook detected.
      before do
      ^^^^^^

IrimieBogdan commented:
Great job with this PR @smortex  and @oanatmaria !
